agent/retry_join.go

```diff
@@ -2,49 +2,45 @@ package agent
 
 import (
 	"fmt"
+	"strings"
 	"time"
+
+	discover "github.com/hashicorp/go-discover"
 )
 
 // RetryJoin is used to handle retrying a join until it succeeds or all
 // retries are exhausted.
 func (a *Agent) retryJoin() {
 	cfg := a.config
-
-	ec2Enabled := cfg.RetryJoinEC2.TagKey != "" && cfg.RetryJoinEC2.TagValue != ""
-	gceEnabled := cfg.RetryJoinGCE.TagValue != ""
-	azureEnabled := cfg.RetryJoinAzure.TagName != "" && cfg.RetryJoinAzure.TagValue != ""
-
-	if len(cfg.RetryJoin) == 0 && !ec2Enabled && !gceEnabled && !azureEnabled {
+	if len(cfg.RetryJoin) == 0 {
 		return
 	}
 
+	// split retry join addresses from go-discover statements
+	var addrs []string
+	var disco string
+	for _, addr := range cfg.RetryJoin {
+		if strings.Contains(addr, "provider=") {
+			disco = addr
```

Comment:
I think this could work for multi-cloud setups but messing this up is a serious operator error. I fail to see how someone can shoot themself in the foot by accident. I can simplify the code a bit  more.